mapper.go

```diff
@@ -107,40 +118,105 @@ func (m *metricMapper) initFromString(fileContents string) error {
 
 	m.mutex.Lock()
 	defer m.mutex.Unlock()
-	m.mappings = parsedMappings
+	m.Mappings = parsedMappings
 
 	mappingsCount.Set(float64(len(parsedMappings)))
 
 	return nil
 }
 
+func (m *metricMapper) initFromYAMLString(fileContents string) error {
+	var n metricMapper
+
+	if err := yaml.Unmarshal([]byte(fileContents), &n); err != nil {
+		return err
+	}
+
+	if n.Defaults.Buckets == nil || len(n.Defaults.Buckets) == 0 {
+		n.Defaults.Buckets = prometheus.DefBuckets
+	}
+
+	for i := range n.Mappings {
+		currentMapping := &n.Mappings[i]
+
+		if !metricLineRE.MatchString(currentMapping.Match) {
+			return fmt.Errorf("invalid match: %s", currentMapping.Match)
+		}
+
+		// check that label is correct
+		for k, v := range currentMapping.Labels {
+			label := fmt.Sprintf("%s=%q", k, v)
+			if len(labelLineRE.FindStringSubmatch(label)) != 3 {
+				return fmt.Errorf("invalid label: %s: %s", k, v)
+			}
+			if k == "name" && !metricNameRE.MatchString(v) {
+				return fmt.Errorf("metric name '%s' doesn't match regex '%s'", v, metricNameRE)
+			}
+		}
+
+		if _, ok := currentMapping.Labels["name"]; !ok {
+			return fmt.Errorf("Line %d: metric mapping didn't set a metric name", i)
```

Comment:
FWIW, errors in `updateMapping` in this same file use uppercase "L".  I'll change my to lower.